clients/drcachesim/tools/record_view.cpp

```diff
@@ -0,0 +1,215 @@
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of Google, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+#include "record_view.h"
+
+#include <cstdio>
+#include <inttypes.h>
+#include <iostream>
+#include <stdint.h>
+#include <string>
+
+#include "memtrace_stream.h"
+#include "trace_entry.h"
+
+namespace dynamorio {
+namespace drmemtrace {
+
+typedef unsigned int uint;
+
+record_analysis_tool_t *
+record_view_tool_create(uint64_t sim_refs)
+{
+    return new dynamorio::drmemtrace::record_view_t(sim_refs);
+}
+
+record_view_t::record_view_t(uint64_t sim_refs)
+    : sim_refs_(sim_refs)
+{
+}
+
+record_view_t::~record_view_t()
+{
+}
+
+bool
+record_view_t::should_skip(void)
+{
+    if (sim_refs_ > 0) {
+        --sim_refs_;
+        return false;
+    }
+    return true;
+}
+
+bool
+record_view_t::parallel_shard_supported()
+{
+    return false;
+}
+
+std::string
+record_view_t::initialize_shard_type(shard_type_t shard_type)
+{
+    return "";
+}
+
+void *
+record_view_t::parallel_shard_init_stream(int shard_index, void *worker_data,
+                                          memtrace_stream_t *shard_stream)
+{
+    return shard_stream;
+}
+
+bool
+record_view_t::parallel_shard_exit(void *shard_data)
+{
+    return true;
+}
+
+std::string
+record_view_t::parallel_shard_error(void *shard_data)
+{
+    return "";
+}
+
+bool
+record_view_t::parallel_shard_memref(void *shard_data, const trace_entry_t &entry)
+{
+    if (should_skip())
+        return true;
+
+    trace_type_t trace_type = (trace_type_t)entry.type;
+    if (trace_type == TRACE_TYPE_INVALID) {
+        std::cerr << "ERROR: trace_entry_t invalid.\n";
+        return false;
+    }
+
+    std::string trace_type_name = std::string(trace_type_names[trace_type]);
+
+    /* Large if-else for all TRACE_TYPE_.  Prints one line per trace_entry_t.
```

Comment:
Replacing this with a switch-case may be more readable.

Reply:
That would make it even larger though, and we'd be basically replicating utility functions like: type_is_intr() or type_has_address().

Reply:
> That would make it even larger though

Do you mean there'd be more lines of code? I'd still suggest that a switch fits the bill better. It makes it clear that we're only doing equality checks. Also: https://stackoverflow.com/questions/427760/when-to-use-if-else-if-else-over-switch-statements-and-vice-versa.

 > we'd be basically replicating utility functions like: type_is_intr() or type_has_address().

Can you clarify this concern? I don't understand.

Reply:
Yes, it would be larger (not my main concern though).
With "replicating utility functions like type_is_instr()" I mean that if we move to a switch we can't use type_is_instr() or type_has_address() anymore, and a case like:
```
...
} else if (type_is_instr(trace_type)) {
...
```
Becomes:
```
switch (...) {
...
case TRACE_TYPE_INSTR:
case TRACE_TYPE_DIRECT_JUMP:
...
case <all remaining TRACE_TYPE_ covered by type_is_instr()>:
...
}
```
Listing all the TRACE_TYPE_ cases in the switch for which type_is_instr() returns `true` is basically replicating what type_is_instr() does, which also means that if in the future we add a new TRACE_TYPE_ that is supposed to represent an instr, then we'd have (at least) two places to update.

I do agree that a switch would look more clear, but I think the cons outweight the pros in this case.
Unless I'm missing something?